services/src/main/java/org/keycloak/services/util/DPoPUtil.java

```diff
@@ -0,0 +1,327 @@
+
+        if (!DPOP_SUPPORTED_ALGS.contains(algorithm)) {
+            throw new VerificationException("Unsupported DPoP algorithm: " + header.getAlgorithm());
+        }
+
+        JWK key = header.getKey();
+
+        if (key == null) {
+            throw new VerificationException("No JWK in DPoP header");
+        } else {
+            KeyWrapper wrapper = JWKSUtils.getKeyWrapper(key);
+            if (wrapper.getPublicKey() == null) {
+                throw new VerificationException("No public key in DPoP header");
+            }
+            if (wrapper.getPrivateKey() != null) {
+                throw new VerificationException("Private key is present in DPoP header");
+            }
+        }
+
+        key.setAlgorithm(header.getAlgorithm().name());
+
+        SignatureVerifierContext signatureVerifier = session.getProvider(SignatureProvider.class, algorithm).verifier(key);
+        verifier.verifierContext(signatureVerifier);
+        verifier.withChecks(
+                DPoPClaimsCheck.INSTANCE,
+                new DPoPHTTPCheck(request, uri),
+                new DPoPIsActiveCheck(session, client),
+                new DPoPReplayCheck(session, client));
+
+        if (tokenString != null) {
+            verifier.withChecks(new DPoPAccessTokenHashCheck(tokenString));
+        }
+
+        try {
+            DPoP dPoP = verifier.verify().getToken();
+            dPoP.setThumbprint(JWKSUtils.computeThumbprint(key));
+            return dPoP;
+        } catch (DPoPVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("DPoP verification failure", ex);
+        }
+
+    }
+
+    public static void validateBinding(AccessToken token, DPoP dPoP) throws VerificationException {
+        try {
+            TokenVerifier.createWithoutSignature(token)
+                    .withChecks(new DPoPUtil.DPoPBindingCheck(dPoP))
+                    .verify();
+        } catch (TokenVerificationException ex) {
+            throw ex;
+        } catch (VerificationException ex) {
+            throw new VerificationException("Token verification failure", ex);
+        }
+    }
+
+    public static void bindToken(AccessToken token, DPoP dPoP) {
+        AccessToken.Confirmation confirmation = token.getConfirmation();
+
+        if (confirmation == null) {
+            confirmation = new AccessToken.Confirmation();
+            token.setConfirmation(confirmation);
+        }
+
+        confirmation.setKeyThumbprint(dPoP.getThumbprint());
+    }
+
+    private static class DPoPClaimsCheck implements TokenVerifier.Predicate<DPoP> {
+
+        static final TokenVerifier.Predicate<DPoP> INSTANCE = new DPoPClaimsCheck();
+
+        @Override
+        public boolean test(DPoP t) throws DPoPVerificationException {
+            Long iat = t.getIat();
+            String jti = t.getId(), htu = t.getHttpUri(), htm = t.getHttpMethod();
```

Comment:
It seems that keycloak does not apply such coding style. How about write it as individual 3 lines?

Reply:
Done